tools/pubsys/src/aws/validate_ssm/mod.rs

```diff
@@ -0,0 +1,732 @@
+                .as_ref()
+                .unwrap_or(&vec![
+                    SsmValidationResultStatus::Correct,
+                    SsmValidationResultStatus::Incorrect,
+                    SsmValidationResultStatus::Missing,
+                    SsmValidationResultStatus::Unexpected,
+                ]),
+        );
+
+        // Write the results as JSON
+        serde_json::to_writer_pretty(
+            &File::create(write_results_path).context(error::WriteValidationResultsSnafu {
+                path: write_results_path,
+            })?,
+            &filtered_results,
+        )
+        .context(error::SerializeValidationResultsSnafu)?;
+    }
+
+    Ok(validation_results)
+}
+
+/// Validates SSM parameters in a single region, based on a HashMap (SsmKey, SsmValue) of expected parameters
+/// and a HashMap (SsmKey, String) of actual retrieved parameters. Returns a HashSet of SsmValidationResult objects.
+pub(crate) fn validate_parameters_in_region(
+    expected_parameters: &HashMap<SsmKey, SsmValue>,
+    actual_parameters: &SsmParameters,
+) -> HashSet<SsmValidationResult> {
+    // Clone the HashMap of actual parameters so items can be removed
+    let mut actual_parameters = actual_parameters.clone();
+    let mut results = HashSet::new();
+
+    // Validate all expected parameters, creating an SsmValidationResult object and
+    // removing the corresponding parameter from `actual_parameters` if found
+    for (ssm_key, ssm_value) in expected_parameters {
+        results.insert(SsmValidationResult::new(
+            ssm_key.name.to_owned(),
+            Some(ssm_value.value.clone()),
+            actual_parameters.get(ssm_key).map(|v| v.to_owned()),
+            ssm_key.region.clone(),
+            Some(ssm_value.ami_id.clone()),
+        ));
+        actual_parameters.remove(ssm_key);
+    }
+
+    // Any remaining parameters in `actual_parameters` were not present in `expected_parameters`
+    // and therefore get the `Unexpected` status
+    for (ssm_key, ssm_value) in actual_parameters {
+        results.insert(SsmValidationResult::new(
+            ssm_key.name.to_owned(),
+            None,
+            Some(ssm_value),
+            ssm_key.region.clone(),
+            None,
+        ));
+    }
+    results
+}
+
+type RegionName = String;
+type AmiId = String;
+type ParameterName = String;
+type ParameterValue = String;
+
+/// Parse the lists of parameters whose paths are in `parameter_lists`. Only parse the parameters
+/// in the regions present in `validation_regions`. Return a HashMap of Region mapped to a HashMap of
+/// the parameters in that region, with each parameter being a mapping of `SsmKey` to `SsmValue`.
+pub(crate) async fn parse_parameter_lists(
+    parameter_lists: Vec<PathBuf>,
+    validation_regions: &[String],
+) -> Result<HashMap<Region, HashMap<SsmKey, SsmValue>>> {
+    let mut parameter_map: HashMap<Region, HashMap<SsmKey, SsmValue>> = HashMap::new();
+    for list in parameter_lists {
```

Comment:
NIT: I would prefer to give it to a more expressive variable name. 
```suggestion
    for parameter_list in parameter_lists {
```